pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1672,6 +1679,10 @@ public void addNewSegment(String tableName, SegmentMetadata segmentMetadata, Str
     }
   }
 
+  private Object getTableUpdaterLock(String offlineTableName) {
+    return _tableUpdaterLocks[offlineTableName.hashCode() % _tableUpdaterLocks.length];
```

Comment:
hash code can be negative. use `& INTEGER_MAX` to get a positive. I fell into the same trap in the realtime part :-)

Reply:
good catch :)